chain/consensus/mir/mine.go

```diff
@@ -0,0 +1,190 @@
+package mir
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	lapi "github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/api/v1api"
+	"github.com/filecoin-project/lotus/chain/consensus/common"
+	"github.com/filecoin-project/lotus/chain/types"
+	mir "github.com/filecoin-project/mir/pkg/types"
+)
+
+func Mine(ctx context.Context, miner address.Address, api v1api.FullNode) error {
+	log.Info("Mir miner started")
+	defer log.Info("Mir miner stopped")
+
+	// TODO: Suppose we want to use Mir in Root and in a subnet at the same time.
+	// Do we need two mir agents for that?
+	mirAgent, err := NewMirAgent(uint64(0))
+	if err != nil {
+		return err
+	}
+	mirErrors := mirAgent.Start(ctx)
+	mirHead := mirAgent.App.ChainNotify
+
+	netName, nerr := api.StateNetworkName(ctx)
+	if nerr != nil {
+		return nerr
+	}
+	subnetID := address.SubnetID(netName)
+
+	log.Infof("Mir miner params: network name - %s, subnet ID - %s", netName, subnetID)
+
+	submitting := time.NewTicker(300 * time.Millisecond)
+	defer submitting.Stop()
+
+	for {
+		base, err := api.ChainHead(ctx)
+		if err != nil {
+			log.Errorw("failed to get the head of chain", "error", err)
+			continue
+		}
+
+		select {
+		case merr := <-mirErrors:
+			return xerrors.Errorf("Mir node error: %w", merr)
+		case <-ctx.Done():
+			log.Debug("Mir miner %s: context closed")
+			return nil
+		case mirBlock := <-mirHead:
+			var msgs []*types.SignedMessage
+			var crossMsgs []*types.Message
+
+			log.Debugf(">>>>> received %d messages in Mir block", len(mirBlock))
+
+			for _, tx := range mirBlock {
+				msg, err := parseTx(tx)
+				if err != nil {
+					log.Error("unable to parse a message:", err)
+					log.Info(msg)
+					continue
+				}
+
+				switch m := msg.(type) {
+				case *types.SignedMessage:
+					msgs = append(msgs, m)
+				case *types.UnverifiedCrossMsg:
+					crossMsgs = append(crossMsgs, m.Msg)
+				default:
+					log.Error("received an unknown message")
+				}
+			}
+
+			log.Infof("[subnet: %s, epoch: %d] try to create a block", subnetID, base.Height()+1)
+			bh, err := api.MinerCreateBlock(ctx, &lapi.BlockTemplate{
+				Miner:            miner,
+				Parents:          base.Key(),
+				BeaconValues:     nil,
+				Ticket:           nil,
+				Epoch:            base.Height() + 1,
+				Timestamp:        uint64(time.Now().Unix()),
+				WinningPoStProof: nil,
+				Messages:         msgs,
+				CrossMessages:    crossMsgs,
+			})
+			if err != nil {
+				log.Errorw("creating block failed", "error", err)
+				continue
+			}
+			if bh == nil {
```

Comment:
Hm but doesn't that mean that we lose one batch of transactions that we've just received from Mir? What happens to the transactions jus read from the channel in the select statement?

Reply:
Yes, correct.
We received transactions from App, tried to create a block, and failed, for example, because one tx was invalid or we were unable to sign the block.
In the current design, these all transactions will not be removed from mempool and will be sent later.